src/multiplex/multiplex.rs

```diff
@@ -304,68 +422,24 @@ impl<S, T, E> Multiplex<S, T>
     fn process_out_body_chunk(&mut self, id: RequestId, chunk: Result<Option<T::BodyOut>, T::Error>) {
         trace!("process out body chunk; id={:?}", id);
 
-        match self.out_bodies.get_mut(&id) {
-            Some(body_sender) => {
-                if body_sender.is_ready {
-                    trace!("   --> sender is ready");
-
-                    // Reverse Result & Option
-                    let chunk = match chunk {
-                        Ok(Some(v)) => Some(Ok(v)),
-                        Ok(None) => None,
-                        Err(e) => Some(Err(e)),
-                    };
-
-                    if let Some(chunk) = chunk {
-                        trace!("   --> send chunk");
-
-                        // Send the chunk
-                        body_sender.sender.send(chunk);
-
-                        // See if the sender is ready
-                        match body_sender.sender.poll_ready() {
-                            Ok(Async::Ready(_)) => {
-                                trace!("   --> ready for more");
-                                // The sender is ready for another message
-                                return;
-                            }
-                            Ok(Async::NotReady) => {
-                                // The sender is not ready for another message
-                                body_sender.is_ready = false;
-                                return;
-                            }
-                            Err(_) => {
-                                // The sender has canceled interest, fall
-                                // through and remove the body sender
-                            }
-                        }
-                    }
-
-                    trace!("   --> end of stream");
-
-                    assert!(body_sender.deque.is_empty());
-                    // Remote has dropped interest in the body. Fall through
-                    // and remove the body sender handle.
-                } else {
-                    trace!("   --> queueing chunk");
-
-                    let frame = match chunk {
-                        Ok(chunk) => Frame::Body(id, chunk),
-                        Err(e) => Frame::Error(id, e),
-                    };
-
-                    body_sender.deque.push(frame);
+        {
+            let exchange = match self.exchanges.get_mut(&id) {
+                Some(v) => v,
+                _ => {
+                    trace!("   --> exchange previously aborted; id={:?}", id);
                     return;
                 }
-            }
-            None => {
-                debug!("interest in body canceled");
+            };
+
+            exchange.send_out_chunk(chunk);
```

Comment:
Currently `send_out_chunk` doesn't flush the `out_deque` if it's full, so it looks like this can be called if there are elements on the `out_deque`? That'd cause the body to get received out of order, right?

Oh wait, reading some more, looks like this is why `out_is_ready` is relevant. I'll comment below...